File: server/session_handler/listen_connections.py
import logging
import os
from threading import Thread

from core import server_settings

from .api_session import APISession, StreamLayer
from .data_channels import TCPChannel, UDPChannel


def starts_session(tcp_channel: TCPChannel, udp_channel: UDPChannel, video_path: str):
    stream_layer = StreamLayer(udp_channel, video_path)
    api_session = APISession(video_path, tcp_channel, stream_layer)
    tcp_channel.write_data(str(os.path.getsize(video_path)))
    api_session.wait_comands()
    return


def create_session(tcp_channel: TCPChannel, client_address: str):
    response = tcp_channel.read_datas()
    udp_port, video_path = response[0]
    logging.info(
        "starting a session with %s, streaming video %s to udp port %s",
        client_address,
        video_path,
        udp_port,
    )
    assert len(response[1:]) == 0
    if not os.path.exists(video_path):
        tcp_channel.write_data("error NotFound")
        return
    udp_channel = UDPChannel(client_address, int(udp_port))
    starts_session(tcp_channel, udp_channel, video_path)
    logging.info("Session with client %s was closed", client_address)


def listen_connections():
    tcp_socket = TCPChannel.create_tcp_socket(
        server_settings.server_ip, server_settings.server_port
    )
    logging.debug(
        "Server listening on %s:%d",
        server_settings.server_ip,
        server_settings.server_port,
    )
    while True:
        client_socket, client_address = tcp_socket.accept()
        logging.debug("Connection from %s:%d", *client_address)
        tcp_channel = TCPChannel(client_socket)
        Thread(target=create_session, args=(tcp_channel, client_address[0])).start()

File: serverjava/src/session_handler/Listen_Connections.java
package session_handler;

import java.io.File;
import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.List;

import session_handler.core.Settings.ServerSettings;
import session_handler.Data_channel.TCPChannel;
import session_handler.Data_channel.UDPChannel;
import session_handler.core.Log;
import session_handler.apiSession.SessionManager;

public class Listen_Connections {

    private static final ServerSettings serverSettings = new ServerSettings();
    private static final Log logger = new Log();

    public static void createSession(TCPChannel tcpChannel, String clientAddress) {
        try {
            List<String[]> response = tcpChannel.readDatas();
            if (response.isEmpty() || response.get(0).length < 2) {
                logger.error("Invalid response format", null);
                tcpChannel.writeData("error InvalidFormat");
                return;
            }

            String udpPort = response.get(0)[0];
            String videoPath = response.get(0)[1];

            logger.info(String.format("Starting a session with %s, streaming video %s to UDP port %s",
                    clientAddress, videoPath, udpPort));

            File videoFile = new File(videoPath);
            if (!videoFile.exists()) {
                tcpChannel.writeData("error NotFound");
                return;
            }

            UDPChannel udpChannel = new UDPChannel(clientAddress, Integer.parseInt(udpPort));
            tcpChannel.writeData(String.valueOf(videoFile.length()));

            SessionManager.startSession(tcpChannel, udpChannel, videoPath);

            logger.info(String.format("Session with client %s was closed", clientAddress));

        } catch (IOException e) {
            logger.error("Error in createSession", e);
        }
    }

    public static void listenConnections() {
        try {
            ServerSocket tcpSocket = TCPChannel.createTcpSocket(
                    serverSettings.getServerIp(), serverSettings.getServerPort(), null);

            logger.debug(String.format("Server listening on %s:%d",
                    serverSettings.getServerIp(), serverSettings.getServerPort()));

            while (true) {
                Socket clientSocket = tcpSocket.accept();
                String clientAddress = clientSocket.getInetAddress().getHostAddress();

                logger.debug(String.format("Connection from %s", clientAddress));

                TCPChannel tcpChannel = new TCPChannel(clientSocket);

                // Create a new thread for each session
                new Thread(() -> createSession(tcpChannel, clientAddress)).start();
            }

        } catch (IOException e) {
            logger.error("Error in listenConnections", e);
        }
    }

    public static void main(String[] args) {
        // Initialize logging with server settings
        Log.initLogging(serverSettings);

        listenConnections();
    }
}